clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -232,19 +232,19 @@ static Value *MakeBinaryAtomicValue(
 
 static Value *EmitNontemporalStore(CodeGenFunction &CGF, const CallExpr *E) {
   Value *Val = CGF.EmitScalarExpr(E->getArg(0));
-  Value *Address = CGF.EmitScalarExpr(E->getArg(1));
+  Address Addr = CGF.EmitPointerWithAlignment(E->getArg(1));
 
   Val = CGF.EmitToMemory(Val, E->getArg(0)->getType());
-  LValue LV = CGF.MakeNaturalAlignAddrLValue(Address, E->getArg(0)->getType());
+  LValue LV = CGF.MakeAddrLValue(Addr, E->getArg(0)->getType());
   LV.setNontemporal(true);
   CGF.EmitStoreOfScalar(Val, LV, false);
   return nullptr;
 }
 
 static Value *EmitNontemporalLoad(CodeGenFunction &CGF, const CallExpr *E) {
-  Value *Address = CGF.EmitScalarExpr(E->getArg(0));
+  Address Addr = CGF.EmitPointerWithAlignment(E->getArg(0));
 
-  LValue LV = CGF.MakeNaturalAlignAddrLValue(Address, E->getType());
+  LValue LV = CGF.MakeAddrLValue(Addr, E->getType());
```

Comment:
Can we land this separately with appropriate tests?  It looks like it changes the behavior (the computed alignment of a pointer can be less than natural alignment, I think).

Reply:
My thinking was that the call to `EmitPointerWithAlignment`, which sets `Addr`, would compute the correct alignment.

Do you have an example that shows the alignment is lower after this change?

Reply:
EmitPointerWithAlignment tries to compute the alignment based on the underlying lvalue.  This can be higher or lower than the natural alignment of the type.  Say you have something like `vec f() { struct S { char c[16]; } x; return __temporal_load((vec*)x.c); }`.  It looks through the cast, sees the field is unaligned, and therefore concludes the pointer is unaligned.  This is arguably an improvement, but it's a significant change to the generated code.

Reply:
Yeah, I agree that (1) this is an improvement and (2) we should probably isolate all of these alignment changes as separate patches.

Reply:
The patch that fixes the alignment: https://github.com/llvm/llvm-project/pull/75675